microcloud/cmd/microcloud/main_init.go

```diff
@@ -414,6 +419,68 @@ func AddPeers(sh *service.Handler, systems map[string]InitSystem) error {
 	return nil
 }
 
+func validateSystems(s *service.Handler, systems map[string]InitSystem) (err error) {
+	curSystem, bootstrap := systems[s.Name]
+	if !bootstrap {
+		return nil
+	}
+
+	// Assume that the UPLINK network on each system is the same, so grab just
+	// the gateways from the current node's UPLINK to verify against the other
+	// systems' management net addrs
+	hasIP4Gateway, hasIP6Gateway := false, false
+	var ip4Gateway, ip6Gateway string
+	var ip4GatewayNet, ip6GatewayNet *net.IPNet
+
+	for _, network := range curSystem.Networks {
+		if network.Type == "physical" && network.Name == "UPLINK" {
+			ip4Gateway, hasIP4Gateway = network.Config["ipv4.gateway"]
+			ip6Gateway, hasIP6Gateway = network.Config["ipv6.gateway"]
+			ovnRanges, hasOVNRanges := network.Config["ipv4.ovn.ranges"]
+
+			if hasIP4Gateway {
+				_, ip4GatewayNet, err = net.ParseCIDR(ip4Gateway)
+				if err != nil {
+					return fmt.Errorf("Invalid ipv4.gateway %q: %w", ip4Gateway, err)
+				}
+			}
+
+			if hasIP6Gateway {
+				_, ip6GatewayNet, err = net.ParseCIDR(ip6Gateway)
+				if err != nil {
+					return fmt.Errorf("Invalid ipv6.gateway %q: %w", ip6Gateway, err)
+				}
+			}
+
+			if hasIP4Gateway && hasOVNRanges {
+				_, err = shared.ParseIPRanges(ovnRanges, ip4GatewayNet)
+				if err != nil {
+					return fmt.Errorf("Invalid ipv4.ovn.ranges %q: %w", ovnRanges, err)
+				}
+			}
+
+			break
+		}
+	}
+
+	for systemName, system := range systems {
+		systemAddr := net.ParseIP(system.ServerInfo.Address)
+		if systemAddr == nil {
+			return fmt.Errorf("Invalid address %q for system %q", system.ServerInfo.Address, systemName)
+		}
+
+		if hasIP4Gateway && ip4GatewayNet.Contains(systemAddr) {
```

Comment:
A comment explaining why this is the case would be great.

Reply:
I'd like to understand more about what this is check is for? Are we truly aiming to check that the system address isn't within the entire uplink subnet, or are we wanting to check something else?
Thanks

Reply:
That's right, we want to check that the system addresses used for internal microcluster traffic don't collide with the uplink gateway subnet. 

The whole uplink subnet may not directly be in use at once (thanks to `ipv4.ovn.ranges`) but the user still did specify that they're reserving that subnet for the uplink network so it shouldn't collide with any of the addresses used for dqlite. 

Additionally I'm not sure how much of the uplink subnet is used for OVN if it's IPv6 as we don't specify a range. So I think in general having a unique subnet per required network interface is the way to go here. 

The corresponding issue mentions a "management subnet" but such a thing doesn't exist in MicroCloud, the subnet chosen for mDNS lookup is more of a filter, and won't necessarily be the smallest subnet that matches the management network. In fact the `/8` subnet used in the example would be impossible to saturate with MicroCloud.

Reply:
> The whole uplink subnet may not directly be in use at once (thanks to `ipv4.ovn.ranges`) but the user still did specify that they're reserving that subnet for the uplink network so it shouldn't collide with any of the addresses used for dqlite.

I'm not sure about this. It might prevent some valid existing setups. Let me try it out and get back to you.

Reply:
Interesting, I was under the impression that the majority of MicroClouds are set up just onnce. Are there many MicroCloud setups that are regularly initialized?

Reply:
You misunderstood. I mean it may prevent setting up previously valid setups.

Reply:
Thinking about this I'm pretty sure having both the microcloud admin network interface and the ovn uplink network interface being connected to the same network (and thus same subnets) is fine and shouldn't be prevented (infact I should think its a likely common setup).

Any ideas where the idea to prevent it came from?

Reply:
It's the second case mentioned in the corresponding isssue #210 

There's a notion of a "management network" that uses the subnet ` 10.42.0.49/8`, and then the uplink gateway collides with it at `10.254.254.1/24`. 

As I mentioned above, there's no such thing as a "management network" in MicroCloud based on a subnet input, rather the lookup subnet just filters the mDNS lookup. But I had thought the extra validation for actually selected systems would be useful. If you think it's too restrictive then we don't need to check each system address against the uplink subnet.

Reply:
I replied here https://github.com/canonical/microcloud/issues/210